src/nbl/asset/utils/CHLSLCompiler.cpp

```diff
@@ -25,22 +25,186 @@
 
-#define TCPP_IMPLEMENTATION
-#include <tcpp/source/tcppLibrary.hpp>
-#undef TCPP_IMPLEMENTATION
-
 using namespace nbl;
 using namespace nbl::asset;
 using Microsoft::WRL::ComPtr;
 
 static constexpr const wchar_t* SHADER_MODEL_PROFILE = L"XX_6_7";
 
+
+
 namespace nbl::asset::hlsl::impl
 {
-    struct DXC {
+    struct DXC 
+    {
         ComPtr<IDxcUtils> m_dxcUtils;
         ComPtr<IDxcCompiler3> m_dxcCompiler;
     };
+
+
+    // for including builtins 
+    struct load_file_or_builtin_to_string
+    {
+
+    template <typename IterContextT>
+    class inner
+    {
+    public:
+        template <typename PositionT>
+        static void init_iterators(IterContextT& iter_ctx, PositionT const& act_pos, boost::wave::language_support language)
+        {
+            using iterator_type = typename IterContextT::iterator_type;
+
+            std::string filepath(iter_ctx.filename.begin(), iter_ctx.filename.end());
+            auto inclFinder = iter_ctx.ctx.get_hooks().m_preprocessOptions.includeFinder;
+            if (inclFinder) 
+            {
+                std::optional<std::string> result;
+                system::path requestingSourceDir(iter_ctx.ctx.get_current_directory().string());
+                if (iter_ctx.type == IterContextT::base_type::file_type::system_header) // is it a sys include (#include <...>)?
+                    result = inclFinder->getIncludeStandard(requestingSourceDir, filepath);
+                else // regular #include "..."
+                    result = inclFinder->getIncludeRelative(requestingSourceDir, filepath);
+
+                if (!result)
+                    BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                        bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                auto& res_str = *result;
+                iter_ctx.instring = res_str;
+            }
+            else // include finder not provided
+            {
+                auto builtin_pair = nbl::builtin::get_resource_runtime(filepath);
+                if (builtin_pair.first) // builtin exists
+                {
+                    iter_ctx.instring = std::string(builtin_pair.first, builtin_pair.first + builtin_pair.second);
+                }
+                else // default boost behavior
+                {
+                    // read in the file
+                    boost::filesystem::ifstream instream(iter_ctx.filename.c_str());
+                    if (!instream.is_open()) {
+                        BOOST_WAVE_THROW_CTX(iter_ctx.ctx, boost::wave::preprocess_exception,
+                            bad_include_file, iter_ctx.filename.c_str(), act_pos);
+                        return;
+                    }
+                    instream.unsetf(std::ios::skipws);
+
+                    iter_ctx.instring.assign(
+                        std::istreambuf_iterator<char>(instream.rdbuf()),
+                        std::istreambuf_iterator<char>());
+                }
+            }
```

Comment:
delete the if-else, you should check that the includeFinder was not null during `locate_include_file`
you can assert it not being null here

Reply:
https://github.com/Devsh-Graphics-Programming/Nabla/pull/608#discussion_r1399347272